x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java

```diff
@@ -244,4 +261,27 @@ public static boolean allocationAllowed(String tierName, Set<DiscoveryNodeRole>
         }
         return false;
     }
+
+    /**
+     * Creates a new DiscoveryNodes object representing the state of the cluster without nodes that are shutting down for removal.
+     * @param nodes the nodes currently present in the cluster
+     * @param nodesShutdownMetadata shutdown metadata, so we know which nodes to filter
+     * @return the nodes in the cluster which are not currently marked for removal
+     */
+    private static DiscoveryNodes filterRemovingNodes(DiscoveryNodes nodes, NodesShutdownMetadata nodesShutdownMetadata) {
+        List<String> removingNodeIds = nodesShutdownMetadata.getAll()
+            .entrySet()
+            .stream()
+            .filter(entry -> entry.getValue().getType().isRemovalType())
+            .map(Map.Entry::getKey)
+            .toList();
+        if (removingNodeIds.isEmpty()) {
+            return nodes;
+        }
+        var builder = DiscoveryNodes.builder(nodes);
+        for (var node : removingNodeIds) {
+            builder.remove(node);
+        }
+        return builder.build();
```

Comment:
Considering https://github.com/elastic/elasticsearch/pull/88436 I wonder if we could avoid having to copy the DiscoveryNodes when calling this allocator with node shutdown in place?

Possibly replace org.elasticsearch.cluster.node.DiscoveryNodes#availableRoles with Map of role -> set of nodeIds and then check (only if there are shutdown entries) that a nodeId exists in the set without corresponding shutdown entry.

Reply:
Ah, that's a good idea, thanks!